distributed/tests/test_client.py

```diff
@@ -7345,6 +7336,109 @@ async def test_dump_cluster_state_json(c, s, a, b, tmp_path, local):
         await c.dump_cluster_state(filename, format="json")
 
 
+def blocked_inc(x, event):
+    event.wait()
+    return x + 1
+
+
+@pytest.mark.parametrize("local", [True, False])
+@pytest.mark.parametrize("_format", ["msgpack", "yaml"])
+@gen_cluster(client=True)
+async def test_inspect_cluster_dump(c, s, a, b, tmp_path, _format, local):
```

Comment:
nit: I'd probably prefer this test in `distributed/tests/test_cluster_dump.py`, since it's not really focused on testing client functionality.

Reply:
Also right now, this is a bit of a monolithic test (you're testing `scheduler_tasks`, `worker_tasks`, `scheduler_story`, `worker_story`, and `split` all in this one function). As we keep adding more cluster-dump inspection tools, I don't think we'd want to keep making this one test longer and longer.

Having to rerun these operations on a cluster to produce the dump is a little annoying (and slow) though. So what if you made a pytest fixture for the dumped state? It could be a module-scoped pytest fixture which created a cluster, ran whatever operations, called `dump_cluster_state`, shut down the cluster, and then returned the dumped state.

```python
@pytest.fixture(scope="module")
@gen_cluster(client=True)  # does this work??
async def dumpfile(c, s, a, b, tmp_path) -> Path:
    filename = tmp_path / "cluster_dump"
    ...
    await c.dump_cluster_state(filename)
    return filename

@pytest.fixture
def dump(dumpfile: Path) -> dict:
    return load_cluster_dump(dumpfile)

def test_scheduler_tasks(dump: dict):
    ...
```

Then you could have unit-style tests of each of the functions without having to rerun a cluster every time to generate the dump.

(Note that I actually often prefer copy-paste reuse in tests over fixtures like this, because if a new test in the future needs to do things a little differently, the barrier feels much higher (either you change the fixture for everything, or you write the one weird test that doesn't use the fixture). So personally I'd be perfectly happy copy-pasting the relevant cluster-dump-generation code for each test too.)

Reply:
Yes, this makes sense.

I've transplanted the test from `test_client.py` into `test_cluster_dump.py` and split into 3 separate tests which test

1. Keys are in the expected place
2. Stories are output
3. `to_yaml` dumps things in the right structure

The `fixtures` were a neat idea and your example worked, for dumpfiles. However I also wanted access to the scheduler, client and workers for reasoning about keys on the cluster. Putting those in a fixture for use by other fixtures and test cases did not work, so I ended up hard-coding the cluster in each test.